### PR TITLE
fix: make `secretKeyRef` a required field in plugins' `configFrom`

### LINKED ISSUE #5061: `+kubebuilder:validation:Optional` is used on types which is not supported by kubebuilder
### Problem statement

KIC's usage of `kubebuilder` core marker: `+kubebuilder:validation:Optional` seems to be inccorect, due to the placement of those markers.

https://book.kubebuilder.io/reference/markers/crd-validation.html mentions that `+kubebuilder:validation:Optional` can be used on fields and packages whereas we use that (somewhat) on types (e.g. https://github.com/Kong/kubernetes-ingress-controller/blob/964f9d264246955a11e4a8bbe778e84c66b09a3c/pkg/apis/configuration/v1/kongclusterplugin_types.go#L32)

This causes the code marker to be interpreted as imposing the validation on package level, which is undesired.

Another consequence of this is that fields in types under that package will have their Go tags understood differently: with `+kubebuilder:validation:Optional`, field that have `json:"name"` tag, will still be marked as optional in the resulting CRD.

Related Kubernetes slack thread: https://kubernetes.slack.com/archives/CAR30FCJZ/p1698841271046759

### Proposed solution

Go through `+kubebuilder:validation:Optional` code markers in CRD types and validate if those should be relocated.

### Acceptance criteria
- [ ] KIC is free of the invalid uses of `+kubebuilder:validation:Optional` described above (perhaps there are more ways to use it incorrectly, but the scope is just this)

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

remove the incorrect usage of annotations `// +kubebuilder:validation:Optional` on type defintions and reserve one `// +kubebuilder:validation:Optional` in packages `v1` and `v1beta1` (in file `docs.go`).

**Which issue this PR fixes**:

fixes #5061 

**Special notes for your reviewer**:

The `// +kubebuilder:validation:Optional`  comments above types are first imported in this PR: https://github.com/Kong/kubernetes-ingress-controller/pull/1757/. I do not know the purpose here, but kept the generated CRDs identical to previous versions. @rainest Do you have more context of that PR? 

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by pmalek** [comment]
How about removing this code marker altogether and make the fields not optional by default. This way, if a field is to be optional we can always add it.

We can make this change:

```diff
diff --git a/pkg/apis/configuration/v1/configsource.go b/pkg/apis/configuration/v1/configsource.go
index 2ba4ea73b..149e0bf58 100644
--- a/pkg/apis/configuration/v1/configsource.go
+++ b/pkg/apis/configuration/v1/configsource.go
@@ -19,10 +19,10 @@ type NamespacedConfigSource struct {
 type SecretValueFromSource struct {
 	// The secret containing the key.
 	// +kubebuilder:validation:Required
-	Secret string `json:"name,omitempty"`
+	Secret string `json:"name"`
 	// The key containing the value.
 	// +kubebuilder:validation:Required
-	Key string `json:"key,omitempty"`
+	Key string `json:"key"`
 }
 
 // NamespacedSecretValueFromSource represents the source of a secret value specifying the secret namespace.
@@ -30,11 +30,11 @@ type SecretValueFromSource struct {
 type NamespacedSecretValueFromSource struct {
 	// The namespace containing the secret.
 	// +kubebuilder:validation:Required
-	Namespace string `json:"namespace,omitempty"`
+	Namespace string `json:"namespace"`
 	// The secret containing the key.
 	// +kubebuilder:validation:Required
-	Secret string `json:"name,omitempty"`
+	Secret string `json:"name"`
 	// The key containing the value.
 	// +kubebuilder:validation:Required
-	Key string `json:"key,omitempty"`
+	Key string `json:"key"`
 }
diff --git a/pkg/apis/configuration/v1/kongclusterplugin_types.go b/pkg/apis/configuration/v1/kongclusterplugin_types.go
index 94e1058ae..2a1d60c95 100644
--- a/pkg/apis/configuration/v1/kongclusterplugin_types.go
+++ b/pkg/apis/configuration/v1/kongclusterplugin_types.go
@@ -29,7 +29,6 @@ import (
 // +kubebuilder:resource:scope=Cluster,shortName=kcp,categories=kong-ingress-controller
 // +kubebuilder:subresource:status
 // +kubebuilder:storageversion
-// +kubebuilder:validation:Optional
 // +kubebuilder:printcolumn:name="Plugin-Type",type=string,JSONPath=`.plugin`,description="Name of the plugin"
 // +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`,description="Age"
 // +kubebuilder:printcolumn:name="Disabled",type=boolean,JSONPath=`.disabled`,description="Indicates if the plugin is disabled",priority=1
@@ -64,7 +63,7 @@ type KongClusterPlugin struct {
 
 	// PluginName is the name of the plugin to which to apply the config.
 	// +kubebuilder:validation:Required
-	PluginName string `json:"plugin,omitempty"`
+	PluginName string `json:"plugin"`
 
 	// RunOn configures the plugin to run on the first or the second or both
 	// nodes in case of a service mesh deployment.
diff --git a/pkg/apis/configuration/v1/kongconsumer_types.go b/pkg/apis/configuration/v1/kongconsumer_types.go
index eb36e3f58..8c04e6594 100644
--- a/pkg/apis/configuration/v1/kongconsumer_types.go
+++ b/pkg/apis/configuration/v1/kongconsumer_types.go
@@ -26,7 +26,6 @@ import (
 // +kubebuilder:subresource:status
 // +kubebuilder:storageversion
 // +kubebuilder:resource:shortName=kc,categories=kong-ingress-controller
-// +kubebuilder:validation:Optional
 // +kubebuilder:printcolumn:name="Username",type=string,JSONPath=`.username`,description="Username of a Kong Consumer"
 // +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`,description="Age"
 // +kubebuilder:printcolumn:name="Programmed",type=string,JSONPath=`.status.conditions[?(@.type=="Programmed")].status`
diff --git a/pkg/apis/configuration/v1/kongingress_types.go b/pkg/apis/configuration/v1/kongingress_types.go
index 3e171b9fc..26b7121cc 100644
--- a/pkg/apis/configuration/v1/kongingress_types.go
+++ b/pkg/apis/configuration/v1/kongingress_types.go
@@ -27,7 +27,6 @@ import (
 // +kubebuilder:subresource:status
 // +kubebuilder:storageversion
 // +kubebuilder:resource:shortName=ki,categories=kong-ingress-controller
-// +kubebuilder:validation:Optional
 // +kubebuilder:validation:XValidation:rule="!has(self.proxy)", message="'proxy' field is no longer supported, use Service's annotations instead"
 // +kubebuilder:validation:XValidation:rule="!has(self.route)", message="'route' field is no longer supported, use Ingress' annotations instead"
 
diff --git a/pkg/apis/configuration/v1/kongplugin_types.go b/pkg/apis/configuration/v1/kongplugin_types.go
index 886f8480e..10eee7bdd 100644
--- a/pkg/apis/configuration/v1/kongplugin_types.go
+++ b/pkg/apis/configuration/v1/kongplugin_types.go
@@ -28,7 +28,6 @@ import (
 // +kubebuilder:subresource:status
 // +kubebuilder:storageversion
 // +kubebuilder:resource:shortName=kp,categories=kong-ingress-controller
-// +kubebuilder:validation:Optional
 // +kubebuilder:printcolumn:name="Plugin-Type",type=string,JSONPath=`.plugin`,description="Name of the plugin"
 // +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`,description="Age"
 // +kubebuilder:printcolumn:name="Disabled",type=boolean,JSONPath=`.disabled`,description="Indicates if the plugin is disabled",priority=1
@@ -64,7 +63,7 @@ type KongPlugin struct {
 
 	// PluginName is the name of the plugin to which to apply the config.
 	// +kubebuilder:validation:Required
-	PluginName string `json:"plugin,omitempty"`
+	PluginName string `json:"plugin"`
 
 	// RunOn configures the plugin to run on the first or the second or both
 	// nodes in case of a service mesh deployment.
```

to not introduce any user facing changes but flip the behavior for developers.

**Comment by pmalek** [comment]
I've created https://github.com/Kong/kubernetes-ingress-controller/issues/5105 to review all CRD level validations and admission webhook code which would be converted into the former.

With change from this PR making a field required is impossible (see the example in #5105). I'd suggest to make the change from https://github.com/Kong/kubernetes-ingress-controller/pull/5103#issuecomment-1798262601, close #5061 and then solve #5105.

**Review by rainest — COMMENTED** [review]
As far as the original PR I likely just misunderstood how the annotation was supposed to be used. AFAICT from re-reading it, the initial versions of our Kubebuilder-based CRDs were lacking validation present in the original CRDs, and that PR added them, e.g. the [original required fields](https://github.com/Kong/kubernetes-ingress-controller/blob/27c29b47e9b752e4279281228c9b3ab5a61cc065/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml#L82-L85) were [not originally present](https://github.com/Kong/kubernetes-ingress-controller/pull/1757/files?show-viewed-files=true&file-filters%5B%5D=#diff-300347e0414f7750ce0ec4175237ac4163810aacaa693d658d05cf324cbc635cR97-R100) when we were migrating to Kubebuilder. The main goal of that PR was to add the required field markers, and I can't recall what, if any purpose there was to the `+kubebuilder:validation:Optional` tags.

The Kubebuilder doc does indicate that the package-level annotation indicates that fields defined by it are optional by default, and that only the fields with the `validation:Required` tag are required, which isn't wrong--the fields without a required tag like `Config` in a Plugin type are indeed optional--but it sounds like that's unnecessary if the JSON serialization hits mark it `omitempty`, and the package-level default introduces other problems.

@pmalek I'm not quite sure I follow the comment. If I read it correctly, we _shouldn't_ make the change proposed here, or at least not until changing the `omitempty`s on the relevant fields? The comment indicates that the change here would actually prevent us from making fields we want to require required, but I'm somewhat confused since the changes here don't actually result in changes to the generated resources (and CI does not indicate they're out of date).

**Comment by pmalek** [comment]
I just tried re-introducing my changes I mentioned here and on last sync and it seems that indeed, it's possible to just slap the `+kubebuilder:validation:Required` on a field and then it will work regardless of the `// +kubebuilder:validation:Optional` being set on a package 🤦 🤷 

```diff
diff --git a/config/crd/bases/configuration.konghq.com_kongconsumers.yaml b/config/crd/bases/configuration.konghq.com_kongconsumers.yaml
index 4e407f191..d1898d2ef 100644
--- a/config/crd/bases/configuration.konghq.com_kongconsumers.yaml
+++ b/config/crd/bases/configuration.konghq.com_kongconsumers.yaml
@@ -151,6 +151,8 @@ spec:
           username:
             description: Username is a Kong cluster-unique username of the consumer.
             type: string
+        required:
+        - username
         type: object
     served: true
     storage: true
diff --git a/pkg/apis/configuration/v1/kongconsumer_types.go b/pkg/apis/configuration/v1/kongconsumer_types.go
index eb36e3f58..47cd8f3ae 100644
--- a/pkg/apis/configuration/v1/kongconsumer_types.go
+++ b/pkg/apis/configuration/v1/kongconsumer_types.go
@@ -37,6 +37,7 @@ type KongConsumer struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// Username is a Kong cluster-unique username of the consumer.
+	// +kubebuilder:validation:Required
 	Username string `json:"username,omitempty"`
 
 	// CustomID is a Kong cluster-unique existing ID for the consumer - useful for mapping
```

Anyway, the change in this PR doesn't introcuce a user facing change but **keeps** the "optional" by default. I'd suggest us go the other way. Just add the optional where is indeed needed and remove it from the package level.

**Comment by randmonkey** [comment]
If you added `omitempty` in JSON tag of a field, the `// +kubebuilder:validation:Required` comment will not take effect, unless you have a `// +kubebuilder:validation:Optional` comment on the package level: https://github.com/kubernetes-sigs/controller-tools/issues/599. 
So I removed the `omitempty` notation in JSON tags of required fields. After the change, these fields of CRDs become required:
`KongConsumer.Username`
`KongPlugin.ConfigFrom.SecretKeyRef` (if `ConfigFrom` is non-empty)
`KongClusterPlugin.ConfigFrom.SecretKeyRef`  (if `ConfigFrom` is non-empty)

**Review by pmalek — COMMENTED** [review]
Apart leaving in `// +kubebuilder:validation:Required` code markers I'd check the `secretKeyRef` change. I'm not sure if making the field required wouldn't cause the enclosing CRD (`KongClusterPlugin` and `KongPlugin`) fail validation if `config` is used (instead of `configFrom` which uses `secretKeyRef` 🤔 

I'd also change the PR title as that's not only about the validation tags now but also about changing the CRD validations.

File: config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml [hunk 1]
```diff
@@ -84,6 +84,8 @@ spec:
                 - name
                 - namespace
                 type: object
+            required:
+            - secretKeyRef
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:88]: Leaving `ConfigFrom` is still valid after the change. This means when `ConfigFrom` is specified, `secretKeyRef` is required(non-empty).

File: pkg/apis/configuration/v1beta1/ingress_rules.go [hunk 2]
```diff
@@ -40,28 +35,22 @@ type IngressRule struct {
...
 	// Specifies the name of the referenced service.
-	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:MinLength=1
 	ServiceName string `json:"serviceName"`
 
 	// Specifies the port of the referenced service.
 	// +kubebuilder:validation:Minimum=1
 	// +kubebuilder:validation:Maximum=65535
 	// +kubebuilder:validation:Format=int32
-	// +kubebuilder:validation:Required
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply pkg/apis/configuration/v1beta1/ingress_rules.go:65]: Added them back.

File: pkg/apis/configuration/v1/kongconsumer_types.go [hunk 3]
```diff
@@ -37,7 +36,7 @@ type KongConsumer struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// Username is a Kong cluster-unique username of the consumer.
-	Username string `json:"username,omitempty"`
+	Username string `json:"username"`
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply pkg/apis/configuration/v1/kongconsumer_types.go]: I changed them back to `omitempty` because `Username` itself is not required: #5137

**Review by pmalek — APPROVED** [review]
I've changed the title of this PR to more reflect what's being changed. Feel free to amend it if you still feel we can improve it.

Leaving this for @randmonkey to decide if he'd like to merge it with the updated title.

**Review by czeslavo — APPROVED** [review]
I believe that by changing `secretKeyRef` to a required field we also do not close doors to a later potential extension of `ConfigFrom` allowing the use of some different source of config, e.g. `configMapKeyRef`. If we'd like to do so in the future, we may then drop the `required` condition and use CEL to require one of `*KeyRef` to be present.

**Pull request merged**